### PR TITLE
Handle complex input and output types

### PR DESCRIPTION
**Pull request by julienp**
For complex inputs and outputs, we need to take care to recursively
rename the schema names to Python names, and the reverse on the way out.

Ref https://github.com/pulumi/pulumi/issues/18366

### CODE REVIEW

**Comment by julienp** [comment]
This change is part of the following stack:

- #18401
    - #18403
        - #18405
            - #18408
                - #18419 ◀

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

File: sdk/python/lib/pulumi/provider/experimental/provider.py [hunk 1]
```diff
@@ -63,32 +63,94 @@ def construct(
...
-            r[mapping[k]] = v
-        return r
+    def get_type_definition(self, prop: PropertyDefinition) -> TypeDefinition:
+        """Gets the type definition for a property with a type reference."""
+        if not prop.ref:
+            raise ValueError(f"property {prop} is not a complex type")
+        name = prop.ref.split(":")[-1]
+        return self._type_defs[name]
+
+    def map_inputs(self, inputs: Inputs, component_def: ComponentDefinition) -> Inputs:
```
Comment (julienp) [inline sdk/python/lib/pulumi/provider/experimental/provider.py:80]: I want to add Python tests here, but I think I first need to make the schema analysis lazy like in Node.js, so we can create a Provider without running the whole analzyer. I'll do that in a follow up PR>

File: tests/integration/component_provider/python/component-provider-host/provider/component.py [hunk 2]
```diff
@@ -18,14 +18,25 @@
 import pulumi
 
 
+class Nested(TypedDict):
+    nested_str: str
+
+
+class Complex(TypedDict):
+    complex_str: pulumi.Input[str]
+    nested: pulumi.Input[Nested]
```
Comment (mikhailshilkov) [inline tests/integration/component_provider/python/component-provider-host/provider/component.py:27]: Suggested test cases:
- A type that recursively references itself
- A recursive loop of type references

These do happen in existing providers and our schema/codegen knows how to handle those.

File: tests/integration/integration_python_test.go [hunk 3]
```diff
@@ -1880,6 +1888,33 @@ func TestPythonComponentProviderGetSchema(t *testing.T) {
...
+	          "type": "string"
+	        },
+	        "nested": {
+	          "$ref": "#/types/provider:index:Nested"
+	        }
+	      },
+	      "type": "object"
+	    },
+	    "provider:index:Nested": {
+	      "properties": {
```
Comment (mikhailshilkov) [inline tests/integration/integration_python_test.go:1906]: Is `"required"` missing? All types seem to have properties required.

File: sdk/python/lib/pulumi/provider/experimental/provider.py [hunk 4]
```diff
@@ -63,32 +63,94 @@ def construct(
...
+        """Maps the input's names from the schema into Python names."""
+        mapped_input: dict[str, Input[Any]] = {}
+        for schema_name, prop in component_def.inputs.items():
+            input_val = inputs.get(schema_name, None)
+            if input_val is None:
```
Comment (mikhailshilkov) [inline sdk/python/lib/pulumi/provider/experimental/provider.py:85]: I don't know if we want to impose required-ness of properties here or leave it to the ComponentResource itself. I expect we can give a better error message here?

File: tests/integration/integration_python_test.go (hunk 3 above)
Comment (julienp) [reply tests/integration/integration_python_test.go]: Yes, good catch! I missed this when building the schema. Fixed in https://github.com/pulumi/pulumi/pull/18419/commits/04f33d7f1ff755b466d68a191ec7c2f90404cb0b

File: sdk/python/lib/pulumi/provider/experimental/provider.py (hunk 4 above)
Comment (julienp) [reply sdk/python/lib/pulumi/provider/experimental/provider.py:85]: Great idea. I'll track this as a separate issue to stop this PR from getting too large https://github.com/pulumi/pulumi/issues/18458

File: tests/integration/component_provider/python/component-provider-host/provider/component.py (hunk 2 above)
Comment (julienp) [reply tests/integration/component_provider/python/component-provider-host/provider/component.py:27]: Fun! I updated the PR for that.

Self recursive works nicely.

For mutually recursive types, it is a bit tricky, because you have to use a `ForwardRef` in Python, which tracks the type by name. I think my solution works reasonably well for common cases, but you could potentially construct types that it fails with right now, mostly if you declare types outside of the global scope, which is possible, but unlikely (I hope?).

File: sdk/python/lib/pulumi/provider/experimental/analyzer.py [hunk 5]
```diff
@@ -180,16 +212,55 @@ def analyze_property(
             raise ValueError("list types not yet implemented")
         elif isinstance(arg, dict):
             raise ValueError("dict types not yet implemented")
+        elif is_forward_ref(arg):
+            name = cast(ForwardRef, arg).__forward_arg__
+            type_def = self.type_definitions.get(name)
+            if type_def:
+                # Forward ref to a type we saw before, return a reference to it.
+                ref = f"#/types/{self.metadata.name}:index:{name}"
```
Comment (mikhailshilkov) [inline sdk/python/lib/pulumi/provider/experimental/analyzer.py:220]: Could a user have two types with the same name in different parts of their app? Will they conflict then? It's probably an okay limitation, but ideally we would catch it and fail with a good error message.

File: tests/integration/component_provider/python/recursive-types/provider/PulumiPlugin.yaml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+runtime:
+  name: python
+  options:
+    toolchain: pip
+    virtualenv: venv
```
Comment (mikhailshilkov) [inline tests/integration/component_provider/python/recursive-types/provider/PulumiPlugin.yaml:5]: Do we need options here?

**Review by mikhailshilkov — APPROVED** [review]
LGTM directionally. I can't say I understand all the pythonisms like forward reference handling, so maybe another review would be useful.

File: tests/integration/component_provider/python/recursive-types/provider/PulumiPlugin.yaml (hunk 6 above)
Comment (julienp) [reply tests/integration/component_provider/python/recursive-types/provider/PulumiPlugin.yaml:5]: No, good call out. These are the defaults that are used anyway, but we tell users to not set options, so we should not do that in our tests either.

File: sdk/python/lib/pulumi/provider/experimental/analyzer.py (hunk 5 above)
Comment (julienp) [reply sdk/python/lib/pulumi/provider/experimental/analyzer.py:254]: 🤯 yes this could happen for both components and complex types. I added detection here https://github.com/pulumi/pulumi/pull/18419/commits/9f4e961d9c042294927fa98424d109809d94617a by tracking each type's module and raising an error if there are duplicates

```
Duplicate type 'MyComponent': orginally defined in 'lib/test/provider/experimental/testdata/duplicate-type/component_a.py' but also found in 'lib/test/provider/experimental/testdata/duplicate-type/component_b.py'
```
